ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,534 @@
+      cache.update(new AccountTwoKeyMap());
+      return;
+    }
+    Map<String, String> accountMap = zNRecord.getMapField(ACCOUNT_METADATA_MAP_KEY);
+    if (accountMap == null) {
+      logger.debug("The ZNRecord={} to read on path={} does not have a simple map with key={}, clearing local cache",
+          zNRecord, pathToFullAccountMetadata, ACCOUNT_METADATA_MAP_KEY);
+      cache.update(new AccountTwoKeyMap());
+      return;
+    }
+    cache.update(new AccountTwoKeyMap(accountMap));
+    logger.trace("Completed fetching full account metadata set from path={}, took time={}ms", pathToFullAccountMetadata,
+        System.currentTimeMillis() - startTimeMs);
+  }
+
+  /**
+   * Checks if there are duplicate accountId or accountName in the given collection of {@link Account}s.
+   *
+   * @param accounts A collection of {@link Account}s to check duplication.
+   * @return {@code true} if there are duplicated accounts in id or name in the given collection of accounts,
+   *                      {@code false} otherwise.
+   */
+  private boolean hasDuplicateAccountIdOrName(Collection<Account> accounts) {
+    if (accounts == null) {
+      return false;
+    }
+    Set<Short> idSet = new HashSet<>();
+    Set<String> nameSet = new HashSet<>();
+    for (Account account : accounts) {
+      if (!idSet.add(account.getId()) || !nameSet.add(account.getName())) {
+        logger.debug("Accounts to update have conflicting id or name. Conflicting accountId={} accountName={}",
+            account.getId(), account.getName());
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Checks a collection of {@link Account}s if there is any conflict between the {@link Account}s to update and the
+   * {@link Account}s in the local cache.
+   *
+   * @param accounts The collection of {@link Account}s to check with the local cache.
+   * @return {@code true} if there is at least one {@link Account} in {@code accounts} conflicts with the {@link Account}s
+   *                      in the cache, {@code false} otherwise.
+   */
+  private boolean hasConflictWithCache(Collection<Account> accounts) {
+    return hasConflictingAccount(accounts, cache.getAccountTwoKeyMap());
+  }
+
+  /**
+   * Checks if there is any {@link Account} in a given collection of {@link Account}s conflicts against any {@link Account}
+   * in a {@link AccountTwoKeyMap}, according to the Javadoc of {@link AccountService}.
+   *
+   * @param accounts The collection of {@link Account}s to check conflict.
+   * @param accountTwoKeyMap A {@link AccountTwoKeyMap} that represents a group of {@link Account}s to check conflict.
+   * @return {@code true} if there is at least one {@link Account} in {@code accountPairs} conflicts with the existing
+   *                      {@link Account}s in {@code accountTwoKeyMap}, {@code false} otherwise.
+   */
+  private boolean hasConflictingAccount(Collection<Account> accounts, AccountTwoKeyMap accountTwoKeyMap) {
```

Comment:
name the first variable accountsToSet, or something,  and maybe name the other one `currentAccounts` to show that we are comparing the first collection against the second set of accounts.